src/madonna/doxydocs/man/man3/usrlib.3
.TH usrlib.h 3 "28 Sep 2000" "madonna" \" -*- nroff -*-
.ad l
.nh
.SH NAME
usrlib.h \- 
.SH SYNOPSIS
.br
.PP
.SS Defines

.in +1c
.ti -1c
.RI "#define \fBINLINE\fR"
.br
.ti -1c
.RI "#define \fBNO_ERROR\fR  0"
.br
.ti -1c
.RI "#define \fBEDOS\fR  1"
.br
.ti -1c
.RI "#define \fBESYS\fR  2"
.br
.ti -1c
.RI "#define \fBENOTMEM\fR  3"
.br
.ti -1c
.RI "#define \fBEWRCFG\fR  4"
.br
.ti -1c
.RI "#define \fBEINPDAT\fR  5"
.br
.ti -1c
.RI "#define \fBEFNFND\fR  6"
.br
.ti -1c
.RI "#define \fBEUNKNOW\fR  7"
.br
.ti -1c
.RI "#define \fBEWRITE\fR  8"
.br
.ti -1c
.RI "#define \fBETASKIMP\fR  9"
.br
.ti -1c
.RI "#define \fBEINDEX\fR  10"
.br
.in -1c
.SS Functions

.in +1c
.ti -1c
.RI "void \fBusrErr\fR (char*,int)"
.br
.ti -1c
.RI "void \fBwarning\fR (char*,int)"
.br
.ti -1c
.RI "void** \fBallocArray2\fR (int xsize,int ysize,int itemSize)"
.br
.ti -1c
.RI "void \fBfreeArray2\fR (int xsize,void **toFree)"
.br
.in -1c
.SH DEFINE DOCUMENTATION
.PP 
.SS #define EDOS  1
.PP
Definition at line 20 of file usrlib.h.
.SS #define EFNFND  6
.PP
Definition at line 25 of file usrlib.h.
.SS #define EINDEX  10
.PP
Definition at line 29 of file usrlib.h.
.SS #define EINPDAT  5
.PP
Definition at line 24 of file usrlib.h.
.SS #define ENOTMEM  3
.PP
Definition at line 22 of file usrlib.h.
.SS #define ESYS  2
.PP
Definition at line 21 of file usrlib.h.
.SS #define ETASKIMP  9
.PP
Definition at line 28 of file usrlib.h.
.SS #define EUNKNOW  7
.PP
Definition at line 26 of file usrlib.h.
.SS #define EWRCFG  4
.PP
Definition at line 23 of file usrlib.h.
.SS #define EWRITE  8
.PP
Definition at line 27 of file usrlib.h.
.SS #define INLINE
.PP
Definition at line 14 of file usrlib.h.
.SS #define NO_ERROR  0
.PP
Definition at line 19 of file usrlib.h.
.SH FUNCTION DOCUMENTATION
.PP 
.SS void ** allocArray2 (int xsize, int ysize, int itemSize)
.PP
Definition at line 60 of file usrlib.C.
.PP
Referenced by \fBImageMap::ImageMap\fR().
.SS void freeArray2 (int xsize, void ** toFree)
.PP
Definition at line 80 of file usrlib.C.
.PP
Referenced by \fBLayMap::findNetId\fR(), \fBImageMap::~ImageMap\fR(), and \fBLayMap::~LayMap\fR().
.SS void usrErr (char * place, int err)
.PP
Definition at line 40 of file usrlib.C.
.PP
Referenced by \fBDivTable::DivTable\fR(), \fBPlane::Plane\fR(), \fBPlcm::createPlane\fR(), \fBPlcm::doTranspAnalysis\fR(), \fBWindow::getDiv\fR(), \fBParserInterface::initImages\fR(), \fBPlcm::placement\fR(), \fBPlcm::prepare\fR(), \fBParserInterface::processFeeds\fR(), \fBPlcm::read\fR(), \fBParserInterface::read\fR(), \fBParserInterface::setRouteOrient\fR(), \fBParserInterface::setTransparency\fR(), and \fBPlcm::write\fR().
.SS void warning (char * place, int err)
.PP
Definition at line 50 of file usrlib.C.
.PP
Referenced by \fBImageMap::addFeeds\fR().
.SH AUTHOR
.PP 
Generated automatically by Doxygen for madonna from the source code.

src/RELEASENOTES

This file is maintained by: Arjan van Genderen
                            e-mail: arjan@cas.et.tudelft.nl

Sep. 22, 1999
-------------

Features compared to earlier distributions:
  
   - Seadali now also runs when it can allocate only a 
     reduced set of colors. (In the past, this caused
     that seadali could not be used at all on some
     screens.)
   - The maximum length of cell, instance, terminal 
     and net names has been increased from 14 to 256.
   - Net names are now used by trout to create labels
     in the layout.  As a result, the extracted circuit
     will have the same net (node) names as the original
     circuit.
   - Introduced an installation script 'ocean_install'.

Source code of this new OCEAN distribution is currently
not available.   The old OCEAN distribution is available 
from ftp://cas.et.tudelft.nl/pub/ocean.old.


Dec. 3, 1999
------------

Source code for compiling the cacd and OCEAN trees is now 
also available.  See the file COMPILATION.

Installation script was renamed oceanInstall.


Dec. 17, 1999
-------------

When the "no routing" option was used with trout, labels
were deleted from the layout.  This has now been fixed.


Jan. 5, 2000
------------

Another bug fix for handling labels during routing.


April 14, 2000
--------------

Fixed the problem that kissis didn't work when 
$SIS/sis_lib/.[sm]isrc are not available.


April 19, 2000
--------------

Fixed problem with directory names with more than 100 
characters.


June 8, 2000
------------

Fixed core dump for trout that appeared under Linux.


Sep. 12, 2000
-------------

- Added tools getldm and layflat.
- Added manual pages of layflat, fish, dofunc, mksls


Sep. 27, 2000
-------------

- Made extension in ICELLS for new cacd release that will use xcontrol 
  instead of macro and device


June 12, 2001
-------------

- Compiled with new version of cacd software.


March 13, 2002
--------------

- Compiled with new version of cacd software.
- Bug fix for seadali for deleting terminal in area.
- Let trout place and route master pins that are not connected to
  any instance.


August 12, 2002
---------------

- Updated celllibs/fishbone/src/nelsis3/template/sls_prototypes/oplib.ext



src/madonna/phil/usrlib.C
/****************************************************************/
/*  USRLIB_CPP  Various user defined functions.                 */
/*                                                              */
/*    Ireneusz Karkowski                                        */
/*    @(#)usrlib.C 1.6 11/23/92 Delft University of Technology 
/****************************************************************/

#include  <iostream.h>
#include <sea_decl.h>

#ifndef   __STDLIB_H
#include   <stdlib.h>
#endif

#ifndef   __USRLIB_H
#include   "usrlib.h"
#endif
#ifndef __MSDOS__
#include <malloc.h>
#else
#include <alloc.h>
#endif
/*-------------------------------------------------------------------------*/
char  *usrErrCodes[]  = {
/*  NO_ERROR    0   */    "",
/*  EDOS        1   */    "Operating System Dummy",
/*  ESYS        2   */    "C++ System Dummy",
/*  ENOTMEM     3   */    "Not enough memory",
/*  EWRCFG      4   */    "Wrong parameter",
/*  EINPDAT     5   */    "Input data error",
/*  EFNFND      6   */    "File not found",
/*  EUNKNOW     7   */    "Unknown error",
/*  EWRITE      8   */    "Can not write data",
/*  ETASKIMP    9   */    "Can not do task",
/*  EINDEX     10   */    "Index out of range"
};


/*-------------------------------------------------------------------------*/
                   void usrErr(char *place,int  err)
/*
  Displays error message to stderr and exits program.
*/
{
  warning(place,err);
  sdfexit(1);
}/* usrerr */

/*-------------------------------------------------------------------------*/
                   void warning(char  *place,int  err)
/*
  Displays warning message to stderr.
*/
{
  cerr << "\n" << place << " : " << usrErrCodes[err] << " !\n";

}/* warning */

//----------------------------------------------------------------------------
               void  **allocArray2(int xsize,int ysize,int itemSize)
//
// allocates two dimensional array of data objects of requested size.
{
  void  **array;

  if( (array = (void**)calloc(xsize,sizeof(void*))) ==NULL)
    usrErr("allocArray2",ENOTMEM);
  
  void **colPtr=array;
  for(int i=0;i<xsize;i++,colPtr++)
  {
    if( (*colPtr = calloc(ysize,itemSize)) ==NULL)
      usrErr("allocArray2",ENOTMEM);
  }
  return array ;

}// allocArray2  //

//----------------------------------------------------------------------------
               void  freeArray2(int xsize,void **toFree)
//
//
{
  void **array=toFree;

  for(int i=0;i<xsize;i++,toFree++)
  {
#ifdef sparc
    free((char *)*toFree);
#else
    free(*toFree);
#endif
  }
#ifdef sparc
  free((char *)array);
#else
  free(array);
#endif

}// freeArray2  //

